contracts/kip103/TreasuryRebalance.sol

```diff
@@ -0,0 +1,519 @@
+     * _retiredAddress is the address of the retired
+     * _approver is the admin of the retiredAddress
+     */
+    function _updateApprover(address _retiredAddress, address _approver)
+        private
+    {
+        uint256 index = getRetiredIndex(_retiredAddress);
+        require(index != type(uint256).max, "Retired not registered");
+        address[] memory approvers = retirees[index].approvers;
+        for (uint256 i = 0; i < approvers.length; i++) {
+            require(approvers[i] != _approver, "Already approved");
+        }
+        retirees[index].approvers.push(_approver);
+        emit Approved(
+            _retiredAddress,
+            _approver,
+            retirees[index].approvers.length
+        );
+    }
+
+    /**
+     * @dev finalizeRegistration sets the status to Registered,
+     *      After this stage, registrations will be restricted.
+     */
+    function finalizeRegistration()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Initialized)
+    {
+        status = Status.Registered;
+        emit StatusChanged(status);
+    }
+
+    /**
+     * @dev finalizeApproval sets the status to Approved,
+     *      After this stage, approvals will be restricted.
+     */
+    function finalizeApproval()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Registered)
+    {
+        require(
+            getTreasuryAmount() < sumOfRetiredBalance(),
+            "treasury amount should be less than the sum of all retired address balances"
+        );
+        _isRetiredsApproved();
+        status = Status.Approved;
+        emit StatusChanged(status);
+    }
+
+    /**
+     * @dev verify if quorom reached for the retired approvals
+     */
+    function _isRetiredsApproved() private view {
+        for (uint256 i = 0; i < retirees.length; i++) {
+            Retired memory retired = retirees[i];
+            bool isContract = isContractAddr(retired.retired);
+            if (isContract) {
+                (address[] memory adminList, uint256 req) = _getState(
+                    retired.retired
+                );
+                if (retired.approvers.length >= req) {
+                    //if min quorom reached, make sure all approvers are still valid
+                    address[] memory approvers = retired.approvers;
+                    uint256 minApprovals = 0;
+                    for (uint256 j = 0; j < approvers.length; j++) {
+                        _validateAdmin(approvers[j], adminList);
+                        minApprovals++;
+                    }
+                    require(
+                        minApprovals >= req,
+                        "min required admins should approve"
+                    );
+                } else {
+                    revert("min required admins should approve");
+                }
+            } else {
+                require(retired.approvers.length == 1, "EOA should approve");
+            }
```

Comment:
Can you make this code a public view function? Klaytn core should validateAdmin at the execution time. If you make this a public function, then core can easily validate the result. If not, the core should implement the same feature retrieving all admin lists for retired and validating the ownership.

Reply:
thanks!